In [4]:
# @title Notebook Version Marker
from datetime import datetime
import pytz

# Change 'UTC' to your local timezone if preferred (e.g., 'US/Eastern', 'Asia/Shanghai')
TIMEZONE = 'UTC'

now = datetime.now(pytz.timezone(TIMEZONE))
version_mark = now.strftime("%Y.%m.%d_%H%M%S")

print("=" * 40)
print(f"SESSION VERSION: {version_mark}")
print(f"START TIME     : {now.strftime('%A, %b %d, %Y - %I:%M:%S %p %Z')}")
print("=" * 40)
print("Exatract D3IL datasets")

SESSION VERSION: 2026.04.11_133842
START TIME     : Saturday, Apr 11, 2026 - 01:38:42 PM UTC
Exatract D3IL datasets


# Mount Google Drive

In [2]:
from google.colab import drive
drive.mount('/content/drive')

# Create a working folder on your Drive
import os
os.makedirs('/content/drive/MyDrive/FMPCC', exist_ok=True)

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


# Clone the FMPCC Repo (to Drive)

In [ ]:
%%bash
cd /content/drive/MyDrive/FMPCC

# Only clone if not already there
if [ ! -d "FM-PCC" ]; then
    git clone --recurse-submodules "https://github.com/ghubliming/FM-PCC.git"
    echo "Cloned successfully."
else
    echo "Repo already exists, pulling latest..."
    cd FM-PCC && git pull
fi

Repo already exists, pulling latest...
Already up to date.


# Install Miniconda into Colab (to get Python 3.10)

In [ ]:
%%bash
# Install Miniconda (gives us conda + Python 3.10 control)
if [ ! -f "/content/miniconda.sh" ]; then
    wget -q https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh -O /content/miniconda.sh
fi
bash /content/miniconda.sh -b -p /content/miniconda3 -u
echo "✓ Miniconda installed"
/content/miniconda3/bin/conda --version

PREFIX=/content/miniconda3
Unpacking bootstrapper...
Unpacking payload...

Installing base environment...

Preparing transaction: ...working... done
Executing transaction: ...working... done
installation finished.
    You currently have a PYTHONPATH environment variable set. This may cause
    unexpected behavior when running the Python interpreter in Miniconda3.
    For best results, please verify that your PYTHONPATH only points to
    directories of packages that are compatible with the Python interpreter
    in Miniconda3: /content/miniconda3
✓ Miniconda installed
conda 26.1.1


# Create conda env with Python 3.10

In [ ]:
%%bash
# Accept Conda Terms of Service for required channels
/content/miniconda3/bin/conda tos accept --override-channels --channel https://repo.anaconda.com/pkgs/main
/content/miniconda3/bin/conda tos accept --override-channels --channel https://repo.anaconda.com/pkgs/r

# Create env exactly as README says
/content/miniconda3/bin/conda create -n FMPCC python=3.10 -y -q
echo "✓ conda env 'FMPCC' created with Python 3.10"
/content/miniconda3/envs/FMPCC/bin/python --version

accepted Terms of Service for https://repo.anaconda.com/pkgs/main
accepted Terms of Service for https://repo.anaconda.com/pkgs/r
Jupyter detected...
2 channel Terms of Service accepted
Channels:
 - defaults
Platform: linux-64
Solving environment: ...working... done

## Package Plan ##

  environment location: /content/miniconda3/envs/FMPCC

  added / updated specs:
    - python=3.10


The following NEW packages will be INSTALLED:

  _libgcc_mutex      pkgs/main/linux-64::_libgcc_mutex-0.1-main 
  _openmp_mutex      pkgs/main/linux-64::_openmp_mutex-5.1-1_gnu 
  bzip2              pkgs/main/linux-64::bzip2-1.0.8-h5eee18b_6 
  ca-certificates    pkgs/main/linux-64::ca-certificates-2025.12.2-h06a4308_0 
  ld_impl_linux-64   pkgs/main/linux-64::ld_impl_linux-64-2.44-h9e0c5a2_3 
  libexpat           pkgs/main/linux-64::libexpat-2.7.4-h7354ed3_0 
  libffi             pkgs/main/linux-64::libffi-3.4.4-h6a678d5_1 
  libgcc             pkgs/main/linux-64::libgcc-15.2.0-h69a1729_7 
  libgcc-ng   

# Install PyTorch (CUDA124)

In [ ]:
%%bash
/content/miniconda3/envs/FMPCC/bin/pip install \
    torch torchvision \
    --index-url https://download.pytorch.org/whl/cu124 -q
echo "✓ PyTorch installed"
/content/miniconda3/envs/FMPCC/bin/python -c \
    "import torch; print('PyTorch:', torch.__version__); print('CUDA:', torch.cuda.is_available())"

✓ PyTorch installed
PyTorch: 2.6.0+cu124
CUDA: False


# Clone D3IL Code

In [ ]:
%%bash
# 1. Ensure the directory exists before jumping in
mkdir -p /content/drive/MyDrive/FMPCC/FM-PCC/

D3IL="/content/drive/MyDrive/FMPCC/FM-PCC/d3il"
PIP="/content/miniconda3/envs/FMPCC/bin/pip"

if [ ! -f "$PIP" ]; then
    echo "✗ Critical Error: Pip not found at $PIP. Check your Miniconda installation."
    exit 1
fi

# 2. check,clone the git repo
if [ ! -d "$D3IL/.git" ]; then
    echo "d3il missing or incomplete — cleaning and cloning..."
    rm -rf "$D3IL"
    git clone https://github.com/ALRhub/d3il.git "$D3IL"
else
    echo "✓ d3il repository detected."
fi

# 3. Use absolute paths for the pip install
$PIP install "$D3IL/environments/d3il"
$PIP install "$D3IL/environments/d3il/envs/gym_avoiding_env"

✓ d3il repository detected.
Processing ./drive/MyDrive/FMPCC/FM-PCC/d3il/environments/d3il
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
  Created wheel for environments.d3il.d3il_sim: filename=environments_d3il_d3il_sim-0.2-py3-none-any.whl size=159418 sha256=1f065187e5eda516cb0fa71cae01ba3ac7d8d0fce2e27097ab4dcc3eabd1e1c2
  Stored in directory: /tmp/pip-ephem-wheel-cache-vqdzs8yr/wheels/c2/2f/ac/2f2eaaf8a5af74e02b2a4afffb7f356f04b0a6d72bded4ba17
Successfully built environments.d3il.d3il_sim
  Attempting uninstall: environments.d3il.d3il_sim
    Found existing installation: environments.d3il.d3il_sim 0.2
    Uninstalling environments.d3il.d3il_sim-0.2:
      Successfully uninstalled environments.d3il.d3il_

# Install D3IL

In [ ]:
%%bash
D3IL_ROOT="/content/drive/MyDrive/FMPCC/FM-PCC/d3il"
PIP="/content/miniconda3/envs/FMPCC/bin/pip"

# Install d3il_sim core (correct subdirectory)
$PIP install -e "$D3IL_ROOT/environments/d3il" -q

# Install gym_avoiding env
$PIP install -e "$D3IL_ROOT/environments/d3il/envs/gym_avoiding_env" -q

echo "✓ D3IL installed"

✓ D3IL installed


# Install All Requirements

In [ ]:
%%bash
PIP="/content/miniconda3/envs/FMPCC/bin/pip"
cd /content/drive/MyDrive/FMPCC/FM-PCC

echo "========================================="
echo " INSTALLING FROM requirements.txt"
echo "========================================="
echo "Using pip: $($PIP --version)"
echo ""

if [ ! -f requirements.txt ]; then
  echo "ERROR: requirements.txt not found in $(pwd)"
  exit 1
fi

echo "Found requirements.txt -- $(wc -l < requirements.txt) packages listed"
echo ""

PASS=0
FAIL=0
SKIP=0
FAILED_PKGS=()

while IFS= read -r line || [ -n "$line" ]; do
  [[ -z "$line" || "$line" == \#* ]] && continue

  PKG="$line"
  printf "  %-40s" "$PKG"

  RESULT=$($PIP install "$PKG" -q 2>&1)
  STATUS=$?

  if [ $STATUS -eq 0 ]; then
    if echo "$RESULT" | grep -q "already satisfied"; then
      echo "[ already installed ]"
      ((SKIP++))
    else
      echo "[ installed ]"
      ((PASS++))
    fi
  else
    echo "[ FAILED ]"
    echo "    └─ $(echo "$RESULT" | tail -3)"
    FAILED_PKGS+=("$PKG")
    ((FAIL++))
  fi

done < requirements.txt

echo ""
echo "========================================="
echo " SUMMARY"
echo "========================================="
printf "  %-25s %d\n" "Newly installed:"  "$PASS"
printf "  %-25s %d\n" "Already installed:" "$SKIP"
printf "  %-25s %d\n" "Failed:"           "$FAIL"

if [ ${#FAILED_PKGS[@]} -gt 0 ]; then
  echo ""
  echo "  Failed packages:"
  for pkg in "${FAILED_PKGS[@]}"; do
    echo "    - $pkg"
  done
  echo ""
  echo "Setup incomplete. Resolve the above before proceeding."
  exit 1
else
  echo ""
  echo "All packages installed successfully."
fi

 INSTALLING FROM requirements.txt
Using pip: pip 26.0.1 from /content/miniconda3/envs/FMPCC/lib/python3.10/site-packages/pip (python 3.10)

Found requirements.txt -- 144 packages listed

  absl-py==2.1.0                          [ installed ]
  accelerate==0.31.0                      [ installed ]
  cachetools==5.3.3                       [ installed ]
  casadi==3.6.5                           [ installed ]
  certifi==2024.6.2                       [ installed ]
  charset-normalizer==3.3.2               [ installed ]
  chex==0.1.86                            [ installed ]
  clarabel==0.9.0                         [ installed ]
  click==8.1.7                            [ installed ]
  cloudpickle==3.0.0                      [ installed ]
  cmeel==0.53.3                           [ installed ]
  cmeel-assimp==5.3.1                     [ installed ]
  cmeel-boost==1.83.0                     [ installed ]
  cmeel-console-bridge==1.0.2.2           [ installed ]
  cmeel-octomap==1.9.8.2     

# Set PYTHONPATH and MuJoCo renderer

In [ ]:
import sys, os

CONDA_PYTHON = "/content/miniconda3/envs/FMPCC/bin/python"
D3IL_ROOT    = "/content/drive/MyDrive/FMPCC/FM-PCC/d3il"
GYM_AV_PATH  = f"{D3IL_ROOT}/environments/d3il/envs/gym_avoiding_env"

# Add paths needed for d3il absolute imports
sys.path.insert(0, D3IL_ROOT)
sys.path.insert(0, GYM_AV_PATH)

os.environ["MUJOCO_GL"]        = "egl"
os.environ["PYOPENGL_PLATFORM"] = "egl"
os.environ["PYTHONPATH"]       = f"{D3IL_ROOT}:{GYM_AV_PATH}:" + os.environ.get("PYTHONPATH","")

print("✓ Paths and env vars set")

✓ Paths and env vars set


# Set Up wandb

In [ ]:
!echo $WANDB_MODE

In [ ]:
# ── 1. Reset wandb completely ────────────────────────────────────
import wandb
wandb.finish()          # close any existing run

# ── 2. Force online mode ─────────────────────────────────────────
import os
os.environ['WANDB_MODE'] = 'online'   # set BEFORE login

# ── 3. Now login interactively ───────────────────────────────────
wandb.login(relogin=True)             # relogin=True forces fresh auth

# ── 4. Init ──────────────────────────────────────────────────────
run = wandb.init(
    project="FMPCC",
    mode="online",        # explicitly set here too
    config={
        "learning_rate": 1e-3,
        "epochs": 10,
    }
)

print(f"Mode: {wandb.run.settings.mode}")  # should print 'online'
print(f"URL:  {run.url}")

wandb: (1) Create a W&B account
wandb: (2) Use an existing W&B account
wandb: (3) Don't visualize my results
wandb: Enter your choice:

 2


wandb: You chose 'Use an existing W&B account'
wandb: Logging into https://api.wandb.ai. (Learn how to deploy a W&B server locally: https://wandb.me/wandb-server)
wandb: Create a new API key at: https://wandb.ai/authorize?ref=models
wandb: Store your API key securely and do not share it.
wandb: Paste your API key and hit enter:

 ··········


wandb: Appending key for api.wandb.ai to your netrc file: /root/.netrc


Mode: online
URL:  https://wandb.ai/llmmail2021-technical-university-of-munich/FMPCC/runs/9kvpe5fo


 # Full verification

In [ ]:
%%bash
PYTHON="/content/miniconda3/envs/FMPCC/bin/python"
D3IL_ROOT="/content/drive/MyDrive/FMPCC/FM-PCC/d3il"
GYM_AV="/content/drive/MyDrive/FMPCC/FM-PCC/d3il/environments/d3il/envs/gym_avoiding_env"

export PYTHONPATH="$D3IL_ROOT:$GYM_AV"
export MUJOCO_GL="egl"

$PYTHON - << 'EOF'
import warnings
warnings.filterwarnings("ignore")
import sys

print(f"Python:  {sys.version}")

import torch
print(f"✓ PyTorch {torch.__version__} | CUDA: {torch.cuda.is_available()}")

import mujoco
print(f"✓ mujoco {mujoco.__version__}")

from environments.d3il import d3il_sim
print("✓ environments.d3il.d3il_sim")

from environments.d3il.envs.gym_avoiding_env.gym_avoiding.envs.avoiding import ObstacleAvoidanceEnv
env = ObstacleAvoidanceEnv(render=False)
env.start()
obs = env.reset()
print(f"✓ ObstacleAvoidanceEnv OK | obs shape: {obs.shape}")

print("")
print("========================================")
print(" ALL CHECKS PASSED — ready to train!")
print("========================================")
EOF

Python:  3.10.20 (main, Mar 11 2026, 17:46:40) [GCC 14.3.0]
✓ PyTorch 2.6.0+cu124 | CUDA: False
✓ mujoco 2.3.7
✓ environments.d3il.d3il_sim
Final IK error (78 iterations):  7.285862274064339e-06
Final IK error (0 iterations):  7.285862274064339e-06
✓ ObstacleAvoidanceEnv OK | obs shape: (2,)

 ALL CHECKS PASSED — ready to train!



pybullet build time: Nov 28 2023 23:45:17


# Handle Gurobi (Not Used for now)

In [ ]:
# After getting license from <https://www.gurobi.com/academia/>
# Upload your gurobi.lic file to Drive, then:
#import os
#os.environ['GRB_LICENSE_FILE'] = '/content/drive/MyDrive/FMPCC/gurobi.lic'

# Dataset download

## Warning, Direct Load the Dataset from the old DPCC folder in Google Drive

In [3]:
%%bash
ZIP="/content/drive/MyDrive/DPCC/dpcc/d3il/environments/dataset/data/dataset.zip"

echo "=== Top-level folders in zip ==="
unzip -l "$ZIP" | awk '{print $4}' | cut -d'/' -f1 | sort -u

echo ""
echo "=== File count per folder ==="
unzip -l "$ZIP" | awk '{print $4}' | cut -d'/' -f1 | sort | uniq -c | sort -rn

echo ""
echo "=== Total uncompressed size ==="
unzip -l "$ZIP" | tail -1

=== Top-level folders in zip ===

----
aligning
avoiding
inserting
Name
pushing
sorting
stacking

=== File count per folder ===
2638779 sorting
1377112 stacking
 392768 aligning
   2002 pushing
    802 inserting
     99 avoiding
      3 
      1 Name
      1 ----

=== Total uncompressed size ===
17689115292                     4411562 files


Example Code and output (Expected)

---

%%bash
ZIP="/content/drive/MyDrive/DPCC/dpcc/d3il/environments/dataset/data/dataset.zip"

echo "=== Top-level folders in zip ==="
unzip -l "$ZIP" | awk '{print $4}' | cut -d'/' -f1 | sort -u

echo ""
echo "=== File count per folder ==="
unzip -l "$ZIP" | awk '{print $4}' | cut -d'/' -f1 | sort | uniq -c | sort -rn

echo ""
echo "=== Total uncompressed size ==="
unzip -l "$ZIP" | tail -1

```
=== Top-level folders in zip ===

----
aligning
avoiding
inserting
Name
pushing
sorting
stacking

=== File count per folder ===
2638779 sorting
1377112 stacking
 392768 aligning
   2002 pushing
    802 inserting
     99 avoiding
      3
      1 Name
      1 ----

=== Total uncompressed size ===
17689115292                     4411562 files

## Only Extract What We Want with Skip the existing files  (e.g. Aligning)

Extracts `EXTRACT_FOLDER` from `dataset.zip` (From DPCC folder) → local Colab disk → `FM-PCC/d3il/environments/dataset/data/` (FMPCC folder). Skips files already on Drive. Live progress every 500 files. Safe to re-run if interrupted.

In [ ]:
import subprocess, os, shutil, zipfile

DATASET_DIR = "/content/drive/MyDrive/DPCC/dpcc/d3il/environments/dataset/data"
TARGET_ROOT = "/content/drive/MyDrive/FMPCC/FM-PCC/d3il/environments/dataset/data"
ZIP_FILE    = f"{DATASET_DIR}/dataset.zip"

EXTRACT_FOLDER = "aligning"   # <-- change this: aligning / stacking / sorting / ...
LOCAL_TMP      = f"/content/{EXTRACT_FOLDER}_tmp"
DEST_FOLDER    = f"{TARGET_ROOT}/{EXTRACT_FOLDER}"

# ---- 1) Extract to local disk with progress ----
print(f"Extracting '{EXTRACT_FOLDER}/' to local disk...")
shutil.rmtree(LOCAL_TMP, ignore_errors=True)
os.makedirs(LOCAL_TMP, exist_ok=True)

with zipfile.ZipFile(ZIP_FILE, 'r') as zf:
    members = [m for m in zf.namelist() if m.startswith(f"{EXTRACT_FOLDER}/") and not m.endswith('/')]
    total_extract = len(members)
    print(f"  Files to extract: {total_extract}")
    for i, member in enumerate(members, 1):
        zf.extract(member, LOCAL_TMP)
        if i % 500 == 0 or i == total_extract:
            print(f"  [extract] {i}/{total_extract}", flush=True)

print(f"✓ Extracted locally: {total_extract} files\n")

# ---- 2) Copy to Drive, skip existing, with progress ----
print("Copying to Drive (skipping existing)...")
src_root = os.path.join(LOCAL_TMP, EXTRACT_FOLDER)
all_files = []
for dirpath, _, filenames in os.walk(src_root):
    for fn in filenames:
        all_files.append(os.path.join(dirpath, fn))

total_copy = len(all_files)
copied = 0
skipped = 0
for i, src_path in enumerate(all_files, 1):
    rel = os.path.relpath(src_path, src_root)
    dst_path = os.path.join(DEST_FOLDER, rel)
    if os.path.exists(dst_path):
        skipped += 1
    else:
        os.makedirs(os.path.dirname(dst_path), exist_ok=True)
        shutil.copy2(src_path, dst_path)
        copied += 1
    if i % 500 == 0 or i == total_copy:
        print(f"  [copy] {i}/{total_copy} | copied={copied} skipped={skipped}", flush=True)

print(f"\n✓ Done — copied {copied}, skipped {skipped}")

# ---- 3) Cleanup ----
shutil.rmtree(LOCAL_TMP, ignore_errors=True)
print("✓ Local tmp cleaned up")

# ---- 4) Verify ----
count = sum(len(f) for _, _, f in os.walk(DEST_FOLDER))
print(f"\n✓ Files on Drive: {count}")

Extracting 'aligning/' to local disk...
  Files to extract: 390762
  [extract] 500/390762
  [extract] 1000/390762
  [extract] 1500/390762
  [extract] 2000/390762
  [extract] 2500/390762
  [extract] 3000/390762
  [extract] 3500/390762
  [extract] 4000/390762
  [extract] 4500/390762
  [extract] 5000/390762
  [extract] 5500/390762
  [extract] 6000/390762
  [extract] 6500/390762
  [extract] 7000/390762
  [extract] 7500/390762
  [extract] 8000/390762
  [extract] 8500/390762
  [extract] 9000/390762
  [extract] 9500/390762
  [extract] 10000/390762
  [extract] 10500/390762
  [extract] 11000/390762
  [extract] 11500/390762
  [extract] 12000/390762
  [extract] 12500/390762
  [extract] 13000/390762
  [extract] 13500/390762
  [extract] 14000/390762
  [extract] 14500/390762
  [extract] 15000/390762
  [extract] 15500/390762
  [extract] 16000/390762
  [extract] 16500/390762
  [extract] 17000/390762
  [extract] 17500/390762
  [extract] 18000/390762
  [extract] 18500/390762
  [extract] 19000/390762
  [

## Only Extract the avoiding (Overwrite Behavior, old, working version)

In [ ]:
%%bash
DATASET_DIR="/content/drive/MyDrive/DPCC/dpcc/d3il/environments/dataset/data" # Notice Here We use the ZIP from DPCC folder
AVOIDING_DATA="/content/drive/MyDrive/FMPCC/FM-PCC/d3il/environments/dataset/data/avoiding/data"
ZIP_FILE="$DATASET_DIR/dataset.zip"

echo "========================================="
echo " D3IL DATASET SETUP"
echo " Source: ALRhub/d3il README.md"
echo " Link: https://drive.google.com/file/d/1SQhbhzV85zf_ltnQ8Cbge2lsSWInxVa8"
echo "========================================="
echo ""

# ── Step 1: Check if avoiding data already extracted ─────────────
if [ -d "$AVOIDING_DATA" ] && [ "$(ls -A $AVOIDING_DATA)" ]; then
    echo "✓ Dataset already extracted at: $AVOIDING_DATA"
    echo "  Files: $(ls $AVOIDING_DATA | wc -l) found"
    echo "  Skipping download and extraction."
    exit 0
fi

# ── Step 2: Check if zip already downloaded ───────────────────────
if [ -f "$ZIP_FILE" ]; then
    echo "✓ Zip already exists: $(du -sh $ZIP_FILE | cut -f1)"
    echo "  Skipping download."
else
    echo "Zip not found. Downloading..."
    mkdir -p "$DATASET_DIR"

    /content/miniconda3/envs/FMPCC/bin/pip install gdown -q
    echo "✓ gdown ready"

    /content/miniconda3/envs/FMPCC/bin/python -m gdown \
        "https://drive.google.com/uc?id=1SQhbhzV85zf_ltnQ8Cbge2lsSWInxVa8" \
        -O "$ZIP_FILE"

    if [ ! -f "$ZIP_FILE" ]; then
        echo "✗ Download failed — zip file not found"
        exit 1
    fi
    echo "✓ Downloaded: $(du -sh $ZIP_FILE | cut -f1)"
fi
echo ""

# ── Step 3: Extract ONLY avoiding/ to local Colab disk (fast) ────
echo "Extracting only avoiding/ to local disk first..."
LOCAL="/content/avoiding_tmp"
mkdir -p "$LOCAL"
unzip -q "$ZIP_FILE" "avoiding/*" -d "$LOCAL"

if [ $? -ne 0 ]; then
    echo "✗ Extraction failed"
    rm -rf "$LOCAL"
    exit 1
fi
echo "✓ Extracted locally: $(find $LOCAL -type f | wc -l) files | $(du -sh $LOCAL | cut -f1)"
echo ""

# ── Step 4: Copy only avoiding/ to Google Drive ───────────────────
echo "Copying avoiding/ to Google Drive..."
mkdir -p "$DATASET_DIR/avoiding"
cp -r "$LOCAL/avoiding/." "/content/drive/MyDrive/FMPCC/FM-PCC/d3il/environments/dataset/data/avoiding/"
echo "✓ Copied to Drive"

# ── Step 5: Cleanup local temp ────────────────────────────────────
rm -rf "$LOCAL"
echo "✓ Local temp cleaned up"
echo ""

# ── Step 6: Final verification ────────────────────────────────────
echo "========================================="
echo " VERIFICATION"
echo "========================================="
if [ -d "$AVOIDING_DATA" ] && [ "$(ls -A $AVOIDING_DATA)" ]; then
    echo "✓ avoiding/data found on Drive"
    echo "  Files: $(ls $AVOIDING_DATA | wc -l)"
    echo "  Sample: $(ls $AVOIDING_DATA | head -3)"
    echo ""
    echo "✓ Dataset ready — you can now run the smoke test."
else
    echo "✗ avoiding/data still not found."
    echo "  Actual structure:"
    find "$DATASET_DIR" -maxdepth 3 | sort | head -20
    exit 1
fi

 D3IL DATASET SETUP
 Source: ALRhub/d3il README.md
 Link: https://drive.google.com/file/d/1SQhbhzV85zf_ltnQ8Cbge2lsSWInxVa8

✓ Zip already exists: 16G
  Skipping download.

Extracting only avoiding/ to local disk first...
✓ Extracted locally: 97 files | 2.6M

Copying avoiding/ to Google Drive...
✓ Copied to Drive
✓ Local temp cleaned up

 VERIFICATION
✓ avoiding/data found on Drive
  Files: 96
  Sample: env_000_00.pkl
env_001_00.pkl
env_002_00.pkl

✓ Dataset ready — you can now run the smoke test.


# Smoke Test Training Cell (Still DPCC, not updated)

```
%%bash
PYTHON="/content/miniconda3/envs/dpcc/bin/python"
PIP="/content/miniconda3/envs/dpcc/bin/pip"
DPCC="/content/drive/MyDrive/DPCC/dpcc"
D3IL_ROOT="$DPCC/d3il"
GYM_AV="$D3IL_ROOT/environments/d3il/envs/gym_avoiding_env"

export PYTHONPATH="$DPCC:$D3IL_ROOT:$GYM_AV"
export MUJOCO_GL="egl"
export WANDB_MODE="offline"
export MPLBACKEND="agg"

cd $DPCC

# Force reinstall gymnasium and minari to resolve potential import issues
$PIP install --force-reinstall gymnasium minari -q

$PYTHON - << 'EOF'
import matplotlib
matplotlib.use('agg')   # force non-interactive backend before any other import

import torch
import diffuser.utils as utils

exp = 'avoiding-d3il'
seeds = [5]

class Parser(utils.Parser):
    dataset: str = exp
    config: str = 'config.' + exp

for seed in seeds:
    args = Parser().parse_args(experiment='diffusion', seed=seed)
    torch.manual_seed(args.seed)

    # Smoke test overrides
    args.n_train_steps     = 10
    args.n_steps_per_epoch = 10
    args.batch_size        = 2

    print(f"Smoke test: seed={seed} | steps={args.n_train_steps} | batch={args.batch_size}")

    dataset_config = utils.Config(
        args.loader,
        savepath=(args.savepath, 'dataset_config.pkl'),
        env=args.dataset,
        horizon=args.horizon,
        normalizer=args.normalizer,
        preprocess_fns=args.preprocess_fns,
        use_padding=args.use_padding,
        max_path_length=args.max_path_length,
        include_returns=args.include_returns,
        returns_scale=args.max_path_length,
        discount=args.discount,
    )
    dataset = dataset_config()
    observation_dim = dataset.observation_dim
    action_dim      = dataset.action_dim
    goal_dim        = dataset.goal_dim

    model_config = utils.Config(
        args.model,
        savepath=(args.savepath, 'model_config.pkl'),
        horizon=args.horizon,
        transition_dim=observation_dim + action_dim,
        cond_dim=observation_dim,
        dim_mults=args.dim_mults,
        returns_condition=args.returns_condition,
        dim=args.dim,
        condition_dropout=args.condition_dropout,
        device=args.device,
    )

    diffusion_config = utils.Config(
        args.diffusion,
        savepath=(args.savepath, 'diffusion_config.pkl'),
        horizon=args.horizon,
        observation_dim=observation_dim,
        action_dim=action_dim,
        goal_dim=goal_dim,
        n_timesteps=args.n_diffusion_steps,
        loss_type=args.loss_type,
        clip_denoised=args.clip_denoised,
        predict_epsilon=args.predict_epsilon,
        action_weight=args.action_weight,
        loss_discount=args.loss_discount,
        returns_condition=args.returns_condition,
        condition_guidance_w=args.condition_guidance_w,
        device=args.device,
    )

    trainer_config = utils.Config(
        utils.Trainer,
        savepath=(args.savepath, 'trainer_config.pkl'),
        train_test_split=args.train_test_split,
        ema_decay=args.ema_decay,
        n_train_steps=args.n_train_steps,
        n_steps_per_epoch=args.n_steps_per_epoch,
        train_batch_size=args.batch_size,
        train_lr=args.learning_rate,
        gradient_accumulate_every=args.gradient_accumulate_every,
        results_folder=args.savepath,
    )

    model     = model_config()
    diffusion = diffusion_config(model)
    trainer   = trainer_config(diffusion, dataset)

    trainer.train()
    print(f"\n✓ Smoke test PASSED — seed {seed} completed {args.n_train_steps} steps")

print("\n========================================")
print(" SMOKE TEST COMPLETE — DPCC code works!")
print("========================================")
EOF
```

 # Train (using the conda env Python)

In [ ]:
import torch
print(torch.cuda.is_available())       # Must be True
print(torch.cuda.get_device_name(0))   # Should show T4 or similar

False


AssertionError: Torch not compiled with CUDA enabled

Warning:

use Real-time Streaming.
Using !python instead of %%bash ensures that as soon as your script prints "Loading model..." or "Epoch 1", you see it immediately.

In [ ]:
import os

for k in ("WANDB_SERVICE", "WANDB__SERVICE"):
    os.environ.pop(k, None)

In [ ]:
import gymnasium, minari
print("gymnasium:", gymnasium.__version__)
print("minari:", minari.__version__)

ModuleNotFoundError: No module named 'minari'

In [ ]:
import os

# Set paths in the Python environment so they persist
os.environ['FMPCC'] = "/content/drive/MyDrive/FMPCC/FM-PCC"
os.environ['PYTHONPATH'] = f"{os.environ['FMPCC']}:{os.environ['FMPCC']}/d3il:{os.environ['FMPCC']}/d3il/environments/d3il/envs/gym_avoiding_env"

# Set GPU and Backend flags
os.environ['MUJOCO_GL'] = "egl"
os.environ['PYOPENGL_PLATFORM'] = "egl"
os.environ['MPLBACKEND'] = "agg"

# Change directory using python so it sticks for the next command
os.chdir(os.environ['FMPCC'])

# Run the script using '!' to get real-time streaming output
!/content/miniconda3/envs/FMPCC/bin/python scripts/train.py --seeds 6 --use-wandb

[ train ] Seed source: cli --seeds
[ train ] Training seeds: [6]
[ train ] Saved seed manifest to: logs/avoiding-d3il/diffusion/H8_K20_Dmodels.GaussianDiffusion/seeds_config.json
wandb: Currently logged in as: llmmail2021 (llmmail2021-technical-university-of-munich). Use `wandb login --relogin` to force relogin
wandb: wandb version 0.25.1 is available!  To upgrade, please run:
wandb:  $ pip install wandb --upgrade
wandb: Tracking run with wandb version 0.17.5
wandb: Run data is saved locally in /content/drive/MyDrive/FMPCC/FM-PCC/wandb/run-20260319_163154-1t4x5uur
wandb: Run `wandb offline` to turn off syncing.
wandb: Syncing run avoiding-d3il-seed-6
wandb: ⭐️ View project at https://wandb.ai/llmmail2021-technical-university-of-munich/fm-pcc-diffusion
wandb: 🚀 View run at https://wandb.ai/llmmail2021-technical-university-of-munich/fm-pcc-diffusion/runs/1t4x5uur
Traceback (most recent call last):
  File "/content/drive/MyDrive/FMPCC/FM-PCC/scripts/train.py", line 220, in <module>
    da

# NEXT CODE not checked

# Set Up Virtual Display for MuJoCo Rendering

In [ ]:
import os
import subprocess

# 1. Start Xvfb (Virtual Framebuffer)
# We check if it's already running to avoid multiple instances
try:
    subprocess.Popen(['Xvfb', ':1', '-screen', '0', '1400x900x24'])
except FileNotFoundError:
    !apt-get install -y xvfb
    subprocess.Popen(['Xvfb', ':1', '-screen', '0', '1400x900x24'])

# 2. Set Environment Variables
os.environ['DISPLAY'] = ':1'
os.environ['MUJOCO_GL'] = 'egl'
os.environ['PYOPENGL_PLATFORM'] = 'egl'

# 3. Verify MuJoCo using your CONDA python
# This ensures we use the version installed in your 'dpcc' environment
!/content/miniconda3/envs/dpcc/bin/python -c "import mujoco; print('MuJoCo version:', mujoco.__version__)"

MuJoCo version: 2.3.7


# Evaluate

In [ ]:
import os

# 1. Path Setup (Colab specific)
FMPCC = "/content/drive/MyDrive/FMPCC/FM-PCC"
D3IL_ROOT = f"{FMPCC}/d3il"
GYM_AV = f"{D3IL_ROOT}/environments/d3il/envs/gym_avoiding_env"
PYTHON_BIN = "/content/miniconda3/envs/FMPCC/bin/python"

# 2. Environment Setup (Matching your CPU logic but for GPU)
os.environ['MUJOCO_GL'] = 'egl'         # Colab GPU uses EGL
os.environ['PYOPENGL_PLATFORM'] = 'egl'
os.environ['WANDB_MODE'] = 'offline'
os.environ['PYTHONPATH'] = f"{FMPCC}:{D3IL_ROOT}:{GYM_AV}"

# 3. Matplotlib Fix (The \"Unset/Reset\" logic from your CPU)
os.environ.pop('MPLBACKEND', None)
os.environ['MPLBACKEND'] = 'agg'

# 4. Execute with Streaming
os.chdir(FMPCC)

print("--- Running Evaluation ---")
!{PYTHON_BIN} scripts/eval.py

print("\n--- Loading Results ---")
!{PYTHON_BIN} scripts/load_results.py

--- Running Evaluation ---

A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.2.6 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "/content/drive/MyDrive/DPCC/dpcc/scripts/eval.py", line 12, in <module>
    from d3il.environments.d3il.envs.gym_avoiding_env.gym_avoiding.envs.avoiding import ObstacleAvoidanceEnv
  File "/content/drive/MyDrive/DPCC/dpcc/d3il/environments/d3il/envs/gym_avoiding_env/gym_avoiding/envs/__init__.py", line 1, in <module>
    from .avoiding import ObstacleAvoidanceEnv
  File "/content/drive/MyDrive/DPCC/dpcc/d3il/environments/d3il/envs/gym_avoiding_env/gym_avoiding/envs/avoiding.py", line 3, i

#  Visualize

In [ ]:
import os

# Define paths
dpcc_dir = "/content/drive/MyDrive/FMPCC/FM-PCC"
d3il_dir = f"{dpcc_dir}/d3il"
gym_dir = f"{d3il_dir}/environments/d3il/envs/gym_avoiding_env"

# Setup environment variables
os.environ['MUJOCO_GL'] = "egl"
os.environ['PYOPENGL_PLATFORM'] = "egl"
os.environ['DISPLAY'] = ":1"
os.environ['MPLBACKEND'] = "agg"
os.environ['PYTHONPATH'] = f"{dpcc_dir}:{d3il_dir}:{gym_dir}"

# Move to the working directory
os.chdir(dpcc_dir)

# Stream the visualization output
!/content/miniconda3/envs/dpcc/bin/python scripts/visualize_data_constraints.py

Halfspace variant: top-right-hard
Number of feasible trajectories: 1/96, 1.04%
Halfspace variant: top-left-hard
Number of feasible trajectories: 0/96, 0.00%
Halfspace variant: both-hard
Number of feasible trajectories: 2/96, 2.08%
/content/drive/MyDrive/DPCC/dpcc/scripts/visualize_data_constraints.py:165: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()
